rules/DowngradePhp70/Rector/FuncCall/DowngradeDirnameLevelsRector.php

```diff
@@ -0,0 +1,157 @@
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+return dirname($path, $levels);
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+$dirnameFunc = function ($path, $levels) {
+    $dir = null;
+    while (--$levels >= 0) {
+        $dir = dirname($dir ?: $path);
+    }
+    return $dir;
+};
+return $dirnameFunc($path, $levels);
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [FuncCall::class];
+    }
+
+    /**
+     * @param FuncCall $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $levelsArg = $this->getLevelsArg($node);
+        if ($levelsArg === null) {
+            return null;
+        }
+
+        $currentStmt = $node->getAttribute(AttributeKey::CURRENT_STATEMENT);
+        $scope = $currentStmt->getAttribute(AttributeKey::SCOPE);
+
+        $funcVariable = new Variable($this->variableNaming->createCountedValueName('dirnameFunc', $scope));
+
+        $closure = $this->createClosure();
+        $exprAssignClosure = $this->createExprAssign($funcVariable, $closure);
+
+        $this->nodesToAddCollector->addNodesBeforeNode([$exprAssignClosure], $node);
+
+        $node->name = $funcVariable;
+
+        return $node;
+    }
+
+    private function getLevelsArg(FuncCall $funcCall): ?Arg
+    {
+        if (! $this->isName($funcCall, self::DIRNAME)) {
+            return null;
+        }
+
+        if (! isset($funcCall->args[1])) {
+            return null;
+        }
+
+        if (! $funcCall->args[1] instanceof Arg) {
+            return null;
+        }
+
+        return $funcCall->args[1];
+    }
+
+    private function createExprAssign(Variable $var, Expr $expr): Expression
+    {
+        return new Expression(new Assign($var, $expr));
+    }
+
+    private function createClosure(): Closure
+    {
+        $dirVariable = new Variable('dir');
+        $pathVariable = new Variable('path');
+        $levelsVariable = new Variable('levels');
+
+        $closure = new Closure();
+        $closure->params = [
+            new Param($pathVariable),
+            new Param($levelsVariable),
+        ];
+        $closure->stmts[] = $this->createExprAssign($dirVariable, $this->nodeFactory->createNull());
+
+        $whileCond = new GreaterOrEqual(new PreDec($levelsVariable), new LNumber(0));
+        $closure->stmts[] = new While_(
+            $whileCond,
+            [$this->createExprAssign(
+                $dirVariable,
+                $this->createDirnameFuncCall(new Arg(new Ternary($dirVariable, null, $pathVariable)))
+            )]
+        );
+        $closure->stmts[] = new Return_($dirVariable);
+
+        return $closure;
+    }
+
+    private function createDirnameFuncCall(Arg $pathArg): FuncCall
+    {
+        return new FuncCall(new Name(self::DIRNAME), [$pathArg]);
+    }
+}
```

Comment:
We use pre-upgrade code for readability. So even downgraded code can be read by humans.
The syntax for these is dirname(dirname(...)).

Could you update it?

Reply:
I've made the change.
There are still cases that I don't handle, like:
```php
$levels = 2;
dirname($path, $levels);
```

Since Rector relies on PHPStan, it's should be possible to do better but I don't know how.
Do you know if there are others rectors doing something like that? I could use them as examples.